src/main/java/com/fortify/fod/Main.java

```diff
@@ -35,85 +35,126 @@ public static void main(String[] args) {
         }
 
-        if(fc.isRemediationScan  && fc.remediationScanPreference != null){
-            System.out.println("Both --r and -rp cannot be used at the same time , --r is deprecated instead we would prefer you to use -rp for selecting Remediation Scan Preference ");
+        if (fc.isRemediationScan && fc.remediationScanPreference != null) {
+            System.err.println("Both --r and -rp cannot be used at the same time , --r is deprecated instead we would prefer you to use -rp for selecting Remediation Scan Preference ");
+            System.exit(1);
+        }
+
+        // Both bsiToken and ReleaseId are not provided - Exit
+        if (fc.bsiToken == null && fc.releaseId == 0) {
+            System.out.print(com.sun.prism.paint.Color.RED);
+            System.err.println("Release ID or BSI token should be required. The Release ID is HIGHLY recommended for CI usage as the BSI token is being sunset in 2020");
             System.exit(1);
         }
 
+        //Both BsiToken and ReleaseId ar provided , call is executed but with a warning message
+        if (fc.bsiToken != null && fc.releaseId != 0) {
+            System.out.println("The BsiToken provided will be ignored and the Scan Settings from the entered Release Id will be used for triggering the current scan.");
+        }
+
         boolean uploadSucceeded;
-        boolean passFailPolicy= true;
+        boolean passFailPolicy = true;
         int triggeredscanId;
+        String portalUrl = null;
+        String apiUrl = null;
+        FodApi fodApi;
         try {
             Proxy proxy = new Proxy(fc.proxy);
-            BsiToken bsiToken = new BsiTokenConverter().convert(fc.bsiToken);
-            if(bsiToken!= null) {
-                FodApi fodApi = new FodApi(bsiToken.getApiUri(), proxy.getProxyUri() == null ? null : proxy, bsiToken.getPortalUri());
-
-                System.out.println("Authenticating");
-
-                // Has username/password
-                String username, password, grantType;
-                if (fc.hasUserCredentials()) {
-                    username = fc.userCredentials.get(0);
-                    password = fc.userCredentials.get(1);
-                    grantType = fodApi.GRANT_TYPE_PASSWORD;
-                    // Has key/secret
+            BsiToken bsiToken = fc.bsiToken != null ? new BsiTokenConverter().convert(fc.bsiToken) : null;
+            if (fc.bsiToken != null) {
+                fodApi = new FodApi(bsiToken.getApiUri(), proxy.getProxyUri() == null ? null : proxy, bsiToken.getPortalUri());
+                apiUrl = bsiToken.getApiUri();
+                portalUrl = bsiToken.getPortalUri();
+            } else {
+                // If user provides Api Uri
+                if (fc.url.contains("api")) {
+                    apiUrl = fc.url;
+                    portalUrl = fc.url.replaceFirst("api.", "");
```

Comment:
I don't think it's a good approach. I avoided it in the Azure plugin by requesting the `portalUrl` separately. Jenkins does it as well.

Reply:
I discussed it with Eric and he is like how can we train user to only provide portal uri .

Reply:
Changed accordingly.